xmtp_mls/src/groups/mls_sync.rs

```diff
@@ -1713,85 +1743,130 @@ fn extract_message_sender(
     client: impl ScopedGroupClient,
-    provider: &XmtpOpenMlsProvider,
-    openmls_group: &mut OpenMlsGroup,
-    intent_data: UpdateGroupMembershipIntentData,
-    signer: impl Signer,
-) -> Result<Option<PublishIntentData>, GroupError> {
-    let extensions: Extensions = openmls_group.extensions().clone();
-
-    let old_group_membership = extract_group_membership(&extensions)?;
-    let new_group_membership = intent_data.apply_to_group_membership(&old_group_membership);
+    provider: &'a XmtpOpenMlsProvider,
+    new_group_membership: &'a GroupMembership,
+    old_group_membership: &'a GroupMembership,
+) -> Result<MembershipDiffWithKeyPackages, GroupError> {
+    let membership_diff = old_group_membership.diff(new_group_membership);
 
-    // Diff the two membership hashmaps getting a list of inboxes that have been added, removed, or updated
-    let membership_diff = old_group_membership.diff(&new_group_membership);
-
-    // Construct a diff of the installations that have been added or removed.
-    // This function goes to the network and fills in any missing Identity Updates
     let installation_diff = client
         .get_installation_diff(
             provider.conn_ref(),
-            &old_group_membership,
-            &new_group_membership,
+            old_group_membership,
+            new_group_membership,
             &membership_diff,
         )
         .await?;
 
-    let mut new_installations: Vec<Installation> = vec![];
-    let mut new_key_packages: Vec<KeyPackage> = vec![];
+    let mut new_installations = Vec::new();
+    let mut new_key_packages = Vec::new();
+    let mut failed_installations = Vec::new();
 
     if !installation_diff.added_installations.is_empty() {
-        let my_installation_id = &client.context().installation_public_key().to_vec();
-        // Go to the network and load the key packages for any new installation
+        let my_installation_id = client.context().installation_public_key().to_vec();
+
         let key_packages = client
             .get_key_packages_for_installation_ids(
                 installation_diff
                     .added_installations
-                    .into_iter()
-                    .filter(|installation| my_installation_id.ne(installation))
+                    .iter()
+                    .filter(|installation| my_installation_id.ne(*installation))
+                    .cloned()
                     .collect(),
             )
             .await?;
 
-        for key_package in key_packages {
-            // Add a proposal to add the member to the local proposal queue
-            new_installations.push(Installation::from_verified_key_package(&key_package));
-            new_key_packages.push(key_package.inner);
+        tracing::info!("trying to validate keypackages");
+
+        #[cfg(any(test, feature = "test-utils"))]
+        let key_packages = {
+            if is_test_mode_upload_malformed_keypackage() {
```

Comment:
i'm not super keen on adding test code directly into our business logic, it seems too easy to make a mistake or accidentally introduce a bug.


This seems like it can be mocked using [mockall](https://docs.rs/mockall/latest/mockall/) ([example](https://github.com/xmtp/libxmtp/blob/656e039222e410062e97ee29ab3f938a4122997b/xmtp_api/src/mls.rs#L395)) by setting up an ApiClientWrapper with the expected malformed key packages. Is there a blocker to using that here?

Reply:
Tried that tbh, not a fan of my current approach and I'm on the same page with you on it. 
Since only one api needs to be mocked and the rest need to work as normal apis, mixing that couldn't spend more time to figure that out. But we can sync on it maybe you have a better idea.